src/servers/src/http/authorize.rs

```diff
@@ -0,0 +1,282 @@
+
+    fn authorize(&mut self, mut request: Request<B>) -> Self::Future {
+        let user_provider = self.user_provider.clone();
+        Box::pin(async move {
+            request.extensions_mut().insert(UserInfo::default());
+            let user_provider = if let Some(user_provider) = &user_provider {
+                user_provider
+            } else {
+                return Ok(request);
+            };
+
+            let (scheme, credential) = match auth_header(&request) {
+                Ok(auth_header) => auth_header,
+                Err(e) => {
+                    error!("failed to get http authorize header, err: {:?}", e);
+                    return Err(unauthorized_resp());
+                }
+            };
+
+            match scheme {
+                AuthScheme::Basic => {
+                    let (username, pwd) = match decode_basic(credential) {
+                        Ok(basic_auth) => basic_auth,
+                        Err(e) => {
+                            error!("failed to decode basic authorize, err: {:?}", e);
+                            return Err(unauthorized_resp());
+                        }
+                    };
+                    match user_provider
+                        .auth(
+                            Identity::UserId(&username, None),
+                            crate::auth::Password::PlainText(pwd.as_bytes()),
+                        )
+                        .await
+                    {
+                        Ok(user_info) => {
+                            request.extensions_mut().insert(user_info);
+                            Ok(request)
+                        }
+                        Err(e) => {
+                            error!("failed to auth, err: {:?}", e);
+                            Err(unauthorized_resp())
+                        }
+                    }
+                }
+            }
+        })
+    }
+}
+
+fn unauthorized_resp<RespBody>() -> Response<RespBody>
+where
+    RespBody: Body + Default,
+{
+    let mut res = Response::new(RespBody::default());
+    *res.status_mut() = StatusCode::UNAUTHORIZED;
+    res
+}
+
+#[derive(Debug)]
+pub enum AuthScheme {
+    Basic,
+}
+
+impl TryFrom<&str> for AuthScheme {
+    type Error = error::Error;
+
+    fn try_from(value: &str) -> Result<Self> {
+        match value.to_lowercase().as_str() {
+            "basic" => Ok(AuthScheme::Basic),
+            other => error::UnsupportedAuthSchemeSnafu { name: other }.fail(),
+        }
+    }
+}
+
+type Credential<'a> = &'a str;
+
+fn auth_header<B>(req: &Request<B>) -> Result<(AuthScheme, Credential)> {
+    let auth_header = req
+        .headers()
+        .get(http::header::AUTHORIZATION)
+        .context(error::NotFoundAuthHeaderSnafu)?
+        .to_str()
+        .context(error::InvisibleASCIISnafu)?;
+
+    let (auth_scheme, encoded_credentials) = auth_header
+        .split_once(' ')
+        .context(error::InvalidAuthorizationHeaderSnafu)?;
```

Comment:
Do you need to "trim" the leading and trailing whitespaces in the `auth_scheme` and `encoded_credentials` here?